src/sentry/ratelimits/cardinality.py

```diff
@@ -377,7 +392,11 @@ def use_quotas(
             # enforce), we can save the redis call entirely.
             return
 
-        with self.client.pipeline(transaction=False) as pipeline:
+        ctx_mgr = (
+            self.client.pipeline(transaction=False) if self.is_redis_cluster else self.client.map()
+        )
```

Comment:
This might be fine but should a layer be built to encapsulate the cluster/rb DAOs? `pipeline()` and `map()` might return similar objects (they both accept `setex/sadd`) but clearly not the same type and I'd be concerned about introducing errors with this logic.

It seems as if these can behave a bit differently. Presumably nothing is happening in `self.client.pipeline(transaction=False)` until `execute()` is called, but `self.client.map()` is executing the redis commands as it goes.

Reply:
@onewland `ctx_mgr` should've been typed `Union[rb.MappingClient, rediscluster.Something]` so i hoped this was typesafe (method access can only happen if the method exists on both union variants), but after getting legibility feedback from nikhar i pulled in another abstraction anyway. I think that's what you meant with DAO too?

Reply:
I was thinking something slightly different but this addressed my concern